engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java

```diff
@@ -1495,6 +1509,11 @@ private void validateThreadLocal(OperatorMeta om) {
       return;
     }
 
+    if (om.getOperator() instanceof Operator.DelayOperator) {
+      String msg = String.format("Locality %s invalid for delay operator %s", Locality.THREAD_LOCAL, om);
+      throw new ValidationException(msg);
```

Comment:
@davidyan74 : The thread_local should be valid till Delay Operator. So even if there is loop, the thread_local will be valid only delay operator as the output of delay operator has to go in next window.